File: modules/core/src/main/java/de/deutschebahn/bahnhoflive/ui/search/StationSearchAdapter.java
/*
 * SPDX-FileCopyrightText: 2020 DB Station&Service AG <bahnhoflive-opensource@deutschebahn.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package de.deutschebahn.bahnhoflive.ui.search;

import android.util.Log;
import android.view.ViewGroup;

import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import de.deutschebahn.bahnhoflive.BaseApplication;
import de.deutschebahn.bahnhoflive.R;
import de.deutschebahn.bahnhoflive.analytics.TrackingManager;
import de.deutschebahn.bahnhoflive.backend.StopPlaceXKt;
import de.deutschebahn.bahnhoflive.backend.db.ris.model.StopPlace;
import de.deutschebahn.bahnhoflive.backend.hafas.model.HafasStation;
import de.deutschebahn.bahnhoflive.persistence.FavoriteStationsStore;
import de.deutschebahn.bahnhoflive.persistence.RecentSearchesStore;
import de.deutschebahn.bahnhoflive.repository.InternalStation;
import de.deutschebahn.bahnhoflive.repository.timetable.TimetableRepository;
import de.deutschebahn.bahnhoflive.ui.ViewHolder;
import de.deutschebahn.bahnhoflive.ui.hub.DbDeparturesViewHolder;
import de.deutschebahn.bahnhoflive.ui.hub.DeparturesViewHolder;
import de.deutschebahn.bahnhoflive.ui.hub.HubViewModel;
import de.deutschebahn.bahnhoflive.ui.station.timetable.TimetableCollectorConnector;
import de.deutschebahn.bahnhoflive.view.SingleSelectionManager;
import kotlin.Unit;
import kotlinx.coroutines.CoroutineScope;

class StationSearchAdapter extends RecyclerView.Adapter<ViewHolder> {

    private final SingleSelectionManager singleSelectionManager = new SingleSelectionManager(this);

    private final FavoriteStationsStore<InternalStation> favoriteDbStationsStore;
    private final FavoriteStationsStore<HafasStation> favoriteHafasStationsStore;

    private final RecentSearchesStore recentSearchesStore;
    private final HubViewModel hubViewModel;
    private final SearchItemPickedListener searchItemPickedListener;

    private List<? extends StopPlace> dbStations;
    private List<HafasStation> hafasStations;
    private List<SearchResult> searchResults;

    private boolean hafasError;
    private boolean dbError;
    private final LifecycleOwner owner;
    private final TrackingManager trackingManager;
    private final CoroutineScope coroutineScope;
    private final TimetableRepository timetableRepository;



    StationSearchAdapter(FragmentActivity context, RecentSearchesStore recentSearchesStore, SearchItemPickedListener searchItemPickedListener,
                         LifecycleOwner owner, TrackingManager trackingManager, CoroutineScope coroutineScope, TimetableRepository timetableRepository) {
        hubViewModel = new ViewModelProvider(context).get(HubViewModel.class);
        this.coroutineScope = coroutineScope;
        this.timetableRepository = timetableRepository;

        this.favoriteDbStationsStore = BaseApplication.get().getApplicationServices().getFavoriteDbStationStore();
        this.favoriteHafasStationsStore = BaseApplication.get().getApplicationServices().getFavoriteHafasStationsStore();
        this.recentSearchesStore = recentSearchesStore;
        this.searchItemPickedListener = searchItemPickedListener;
        this.owner = owner;
        this.trackingManager = trackingManager;
        singleSelectionManager.addListener(selectionManager -> {
            final SearchResult selectedItem = selectionManager.getSelectedItem(searchResults);
            if (selectedItem instanceof StoredStationSearchResult) {

                final StoredStationSearchResult searchResult = (StoredStationSearchResult) selectedItem;
Log.d("cr", "selItem: " + selectedItem.toString() + " searcgResult: " + searchResult.toString());
                ((StoredStationSearchResult) selectedItem).getTimetable().getFirst().loadIfNecessary();

                if(searchResult.timetableCollectorConnector == null)
                    searchResult.timetableCollectorConnector = new TimetableCollectorConnector(owner);

                // get Station-Abfahrten
                if (searchResult.timetableCollectorConnector != null)
                    searchResult.timetableCollectorConnector.setStationAndRequestDestinationStations(searchResult.station, timetable -> {
                                notifyDataSetChanged();
                                return Unit.INSTANCE;
                            },
                            integer -> {
                                return Unit.INSTANCE;
                            },
                            aBoolean -> {
                                return Unit.INSTANCE;
                            }

                    );
            } else if (selectedItem instanceof StopPlaceSearchResult) { // long click on (temp.) searchresult
                StopPlaceSearchResult searchResult = ((StopPlaceSearchResult) selectedItem);


                // get Station-Abfahrten
//                if(searchResult.getTimetableCollectorConnector() != null)
//                    searchResult.getTimetableCollectorConnector().setStationAndRequestDestinationStations(searchResult.getStation(), timetable -> {
//                                notifyDataSetChanged();
//                                return Unit.INSTANCE;
//                            },
//                            integer -> {
//                                return Unit.INSTANCE;
//                            },
//                            aBoolean -> {
//                                return Unit.INSTANCE;
//                            }
//
//                    );

            } else if (selectedItem instanceof HafasStationSearchResult) {
                ((HafasStationSearchResult) selectedItem).getTimetable().requestTimetable(true, "search");
            }
        });
        showRecents(coroutineScope, timetableRepository);
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        switch (viewType) {
            case 0:
            case 2:
                return new DbDeparturesViewHolder(parent, singleSelectionManager, owner,
                        trackingManager, searchItemPickedListener, TrackingManager.UiElement.ABFAHRT_SUCHE_BHF);
            case 1:
                return new DeparturesViewHolder(parent, owner, singleSelectionManager, trackingManager,
                        searchItemPickedListener, TrackingManager.UiElement.ABFAHRT_SUCHE_OPNV);
            default:
                return new StationSearchViewHolder(parent, R.layout.card_station_suggestion);
        }
    }

    @Override
    public int getItemViewType(int position) {
        final SearchResult searchResult = searchResults.get(position);

        if (searchResult instanceof StoredStationSearchResult) {
            return 0;
        }

        if (searchResult instanceof HafasStationSearchResult) {
            return 1;
        }

        if (searchResult instanceof StopPlaceSearchResult) {
            return 2;
        }
        return 3;
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        final SearchResult searchResult = searchResults.get(position);
        holder.bind(searchResult);
    }

    @Override
    public int getItemCount() {
        return searchResults.size();
    }

    public void setDBStations(final List<? extends StopPlace> stations) {
        dbError = false;
        this.dbStations = stations;
        updateItems();
    }

    private void updateItems() {
        singleSelectionManager.clearSelection();

        searchResults.clear();

        if (dbStations != null) {
            for (StopPlace dbStation : dbStations) {
                final SearchResult searchResult;
                if (dbStation.isDbStation()) {

//                    TimetableCollectorConnector timetableCollectorConnector  = new TimetableCollectorConnector(this.owner);

                    searchResult = new StopPlaceSearchResult(coroutineScope, dbStation,
                            recentSearchesStore, favoriteDbStationsStore, timetableRepository); //, timetableCollectorConnector);
                } else {
                    final HafasStation hafasStation = StopPlaceXKt.toHafasStation(dbStation);
                    searchResult = new HafasStationSearchResult(hafasStation, recentSearchesStore, favoriteHafasStationsStore);
                }
                searchResults.add(searchResult);
            }
        }

        if (hafasStations != null) {
            for (HafasStation hafasStation : hafasStations) {
                searchResults.add(new HafasStationSearchResult(hafasStation, recentSearchesStore, favoriteHafasStationsStore));
            }
        }

        notifyDataSetChanged();
    }

    public void setHafasStations(List<HafasStation> stations) {
        hafasStations = stations;
        hafasError = false;

        updateItems();
    }

    public void showRecents(CoroutineScope coroutineScope, TimetableRepository timetableRepository) {
        singleSelectionManager.clearSelection();

        searchResults = recentSearchesStore.loadRecentStations(coroutineScope, timetableRepository);

        notifyDataSetChanged();
    }

    public void setHafasError() {
        hafasError = true;

        updateItems();
    }

    public void setDBError() {
        dbError = true;

        updateItems();
    }

    public boolean hasErrors() {
        return dbError || hafasError;
    }

    public void clearSelection() {
        singleSelectionManager.clearSelection();
    }
}

File: modules/core/src/main/java/de/deutschebahn/bahnhoflive/ui/search/StationSearchAdapter.kt
/*
 * SPDX-FileCopyrightText: 2020 DB Station&Service AG <bahnhoflive-opensource@deutschebahn.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */
package de.deutschebahn.bahnhoflive.ui.search

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.RecyclerView
import de.deutschebahn.bahnhoflive.BaseApplication.Companion.get
import de.deutschebahn.bahnhoflive.R
import de.deutschebahn.bahnhoflive.analytics.TrackingManager
import de.deutschebahn.bahnhoflive.backend.db.ris.model.StopPlace
import de.deutschebahn.bahnhoflive.backend.hafas.model.HafasStation
import de.deutschebahn.bahnhoflive.backend.toHafasStation
import de.deutschebahn.bahnhoflive.persistence.FavoriteStationsStore
import de.deutschebahn.bahnhoflive.persistence.RecentSearchesStore
import de.deutschebahn.bahnhoflive.repository.InternalStation
import de.deutschebahn.bahnhoflive.repository.timetable.OnStartStopCyclicLoadingOfTimetableListener
import de.deutschebahn.bahnhoflive.repository.timetable.TimetableCollector
import de.deutschebahn.bahnhoflive.repository.timetable.TimetableRepository
import de.deutschebahn.bahnhoflive.ui.ViewHolder
import de.deutschebahn.bahnhoflive.ui.hub.DbDeparturesViewHolder
import de.deutschebahn.bahnhoflive.ui.hub.DeparturesViewHolder
import de.deutschebahn.bahnhoflive.ui.hub.HubViewModel
import de.deutschebahn.bahnhoflive.view.SingleSelectionManager
import kotlinx.coroutines.CoroutineScope

internal class StationSearchAdapter(
    context: FragmentActivity,
    private val recentSearchesStore: RecentSearchesStore,
    private val searchItemPickedListener: SearchItemPickedListener,
    private val owner: LifecycleOwner,
    private val trackingManager: TrackingManager,
    private val coroutineScope: CoroutineScope,
    private val timetableRepository: TimetableRepository,
    onStartOrStopCyclicLoadingOfTimetableListener: OnStartStopCyclicLoadingOfTimetableListener
) : RecyclerView.Adapter<ViewHolder<*>>() {
    private val singleSelectionManager = SingleSelectionManager(this)
    private val favoriteDbStationsStore: FavoriteStationsStore<InternalStation>
    private val favoriteHafasStationsStore: FavoriteStationsStore<HafasStation>
    private val hubViewModel: HubViewModel

    private var dbStations: List<StopPlace>? = null
//    private var hafasStations: List<HafasStation>? = null

    private var searchResults: MutableList<SearchResult> = mutableListOf()

    private var hafasError = false
    private var dbError = false

    init {
        hubViewModel = ViewModelProvider(context)[HubViewModel::class.java]
        favoriteDbStationsStore = get().applicationServices.favoriteDbStationStore
        favoriteHafasStationsStore = get().applicationServices.favoriteHafasStationsStore
        singleSelectionManager.addListener { selectionManager: SingleSelectionManager ->
            val selectedItem = selectionManager.getSelectedItem(searchResults)

            // long click
            val selection = selectionManager.selection
            if (selection == SingleSelectionManager.INVALID_SELECTION) {
                // selection wurde eingeklappt
                onStartOrStopCyclicLoadingOfTimetableListener.onStartStopCyclicLoading(
                    null, null, selection
                ) // permanentes laden abbrechen
            } else if (selectedItem is StoredStationSearchResult) {
                // long click on stored item
                onStartOrStopCyclicLoadingOfTimetableListener.onStartStopCyclicLoading(
                    selectedItem.timetable,
                    null,
                    selection
                )
            } else if (selectedItem is StopPlaceSearchResult) {
                // long click on typed searchresult
                onStartOrStopCyclicLoadingOfTimetableListener.onStartStopCyclicLoading(
                    selectedItem.timetable,
                    null,
                    selection
                )
            } else if (selectedItem is HafasStationSearchResult) {
                selectedItem.timetable.requestTimetable(true, "search", true)
                onStartOrStopCyclicLoadingOfTimetableListener.onStartStopCyclicLoading(
                    null,
                    selectedItem.timetable,
                    selection
                )
            }
        }
        showRecents(coroutineScope, timetableRepository)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder<*> {
        return when (viewType) {
            0, 2 -> DbDeparturesViewHolder(
                LayoutInflater.from(parent.context)
                    .inflate(R.layout.card_departures, parent, false),
                owner,
                singleSelectionManager,
                trackingManager,
                searchItemPickedListener,
                TrackingManager.UiElement.ABFAHRT_SUCHE_BHF
            )

            1 -> DeparturesViewHolder(
                LayoutInflater.from(parent.context)
                    .inflate(R.layout.card_departures, parent, false),
                owner, singleSelectionManager, trackingManager,
                searchItemPickedListener, TrackingManager.UiElement.ABFAHRT_SUCHE_OPNV
            )

            else -> StationSearchViewHolder(
                LayoutInflater.from(parent.context)
                    .inflate(R.layout.card_station_suggestion, parent, false)
            )
        }
    }

    override fun getItemViewType(position: Int): Int {
        val searchResult = searchResults[position]
        if (searchResult is StoredStationSearchResult) {
            return 0
        }
        if (searchResult is HafasStationSearchResult) {
            return 1
        }
        return if (searchResult is StopPlaceSearchResult) {
            2
        } else 3
    }

    override fun onBindViewHolder(holder: ViewHolder<*>, position: Int) {
        val searchResult = searchResults[position]

        when{
            (searchResult is StoredStationSearchResult)->
                (holder as DbDeparturesViewHolder).bind(searchResult)

            (searchResult is HafasStationSearchResult)->
                (holder as DeparturesViewHolder).bind(searchResult)

            (searchResult is StopPlaceSearchResult)-> {
                @Suppress("Unchecked_Cast")
                (holder as DbDeparturesViewHolder).bind(searchResult as StationSearchResult<InternalStation?, TimetableCollector?>)
            }
            else ->
              (holder as StationSearchViewHolder).bind(searchResult)
        }


    }

    override fun getItemCount(): Int {
        return searchResults.size
    }



    private fun updateItems() {
        singleSelectionManager.clearSelection()
        searchResults.clear()

        dbStations?.let {
            for (dbStation in it) {
                val searchResult: SearchResult = if (dbStation.isDbStation) {
                    StopPlaceSearchResult(
                        coroutineScope, dbStation,
                        recentSearchesStore, favoriteDbStationsStore, timetableRepository
                    )
                } else {
                    val hafasStation = dbStation.toHafasStation()
                    HafasStationSearchResult(
                        hafasStation,
                        recentSearchesStore,
                        favoriteHafasStationsStore
                    )
                }
                searchResults.add(searchResult)
            }
        }

//        hafasStations?.let {
//            for (hafasStation in it) {
//                searchResults.add(
//                    HafasStationSearchResult(
//                        hafasStation,
//                        recentSearchesStore,
//                        favoriteHafasStationsStore
//                    )
//                )
//            }
//        }
        notifyDataSetChanged()
    }

    fun showRecents(coroutineScope: CoroutineScope?, timetableRepository: TimetableRepository?) {
        singleSelectionManager.clearSelection()
        searchResults.clear()
        searchResults.addAll(recentSearchesStore.loadRecentStations(coroutineScope, timetableRepository!!))

//            recentSearchesStore.loadRecentStations(coroutineScope, timetableRepository!!)
        notifyDataSetChanged()
    }


//    fun setHafasStations(stations: List<HafasStation>?) {
//        hafasStations = stations
//        hafasError = false
//        updateItems()
//    }
//
//    fun setHafasError() {
//        hafasError = true
//        updateItems()
//    }

    fun setDBStations(stations: List<StopPlace>?) {
        dbError = false
        dbStations = stations
        updateItems()
    }

    fun setDBError() {
        dbError = true
        updateItems()
    }

    fun hasErrors(): Boolean {
        return dbError || hafasError
    }

    fun clearSelection() {
        singleSelectionManager.clearSelection()
    }
}
